src/integrations/terminal/Terminal.ts

```diff
@@ -0,0 +1,228 @@
+import * as vscode from "vscode"
+import pWaitFor from "p-wait-for"
+import { ExitCodeDetails, mergePromise, TerminalProcess, TerminalProcessResultPromise } from "./TerminalProcess"
+
+export class Terminal {
+	public terminal: vscode.Terminal
+	public busy: boolean
+	public id: number
+	private stream?: AsyncIterable<string>
+	public running: boolean
+	private streamClosed: boolean
+	public process?: TerminalProcess
+	public taskId?: string
+	public completedProcesses: TerminalProcess[] = []
+	private initialCwd: string
+
+	constructor(id: number, terminal: vscode.Terminal, cwd: string) {
+		this.id = id
+		this.terminal = terminal
+		this.busy = false
+		this.running = false
+		this.streamClosed = false
+
+		// Initial working directory is used as a fallback when
+		// shell integration is not yet initialized or unavailable:
+		this.initialCwd = cwd
+	}
+
+	/**
+	 * Gets the current working directory from shell integration or falls back to initial cwd
+	 * @returns The current working directory
+	 */
+	public getCurrentWorkingDirectory(): string {
+		// Try to get the cwd from shell integration if available
+		if (this.terminal.shellIntegration?.cwd) {
+			return this.terminal.shellIntegration.cwd.fsPath
+		} else {
+			// Fall back to the initial cwd
+			return this.initialCwd
+		}
+	}
+
+	/**
+	 * Gets the terminal's stream
+	 */
+	public getStream(): AsyncIterable<string> | undefined {
+		return this.stream
+	}
+
+	/**
+	 * Checks if the stream is closed
+	 */
+	public isStreamClosed(): boolean {
+		return this.streamClosed
+	}
+
+	/**
+	 * Sets the active stream for this terminal and notifies the process
+	 * @param stream The stream to set, or undefined to clean up
+	 * @throws Error if process is undefined when a stream is provided
+	 */
+	public setActiveStream(stream: AsyncIterable<string> | undefined): void {
+		this.stream = stream
```

Comment:
I'm curious why we are storing a reference to the stream here; it doesn't seem to be used anywhere. I supposed it could be used in `TerminalProcess#run`, but we use the reference passed in the event.